src/compiler/checker.ts

```diff
@@ -22168,8 +22168,17 @@ namespace ts {
         // constituent types keyed by the literal types of the property by that name in each constituent type.
         function getKeyPropertyName(unionType: UnionType): __String | undefined {
             const types = unionType.types;
-            // We only construct maps for large unions with non-primitive constituents.
-            if (types.length < 10 || getObjectFlags(unionType) & ObjectFlags.PrimitiveUnion) {
+            // We only construct maps for unions with many non-primitive constituents.

```

Comment:
I'd suggest using `countWhere`, and count the number of types for which we'll actually query properties:

```suggestion
            // We only construct maps for unions with many non-primitive constituents.
            if (types.length < 10 || getObjectFlags(unionType) & ObjectFlags.PrimitiveUnion ||
                countWhere(types, t => !!(t.flags & (TypeFlags.Object | TypeFlags.InstantiableNonPrimitive))) < 10) {
```

Reply:
Oh, I had overlooked that below flags expression was what I want. Thank you!

```ts
                const keyPropertyName = forEach(types, t =>
                    t.flags & (TypeFlags.Object | TypeFlags.InstantiableNonPrimitive) ?
                        forEach(getPropertiesOfType(t), p => isUnitType(getTypeOfSymbol(p)) ? p.escapedName : undefined) :
                        undefined);
```